websites/A/AnimePill/presence.ts

```diff
@@ -0,0 +1,147 @@
+interface PageContext {
+  middleware: (ref: Window, ...args: unknown[]) => boolean;
+  exec: (
+    context: Presence,
+    data: PresenceData,
+    options: { [key: string]: unknown }
+  ) => Promise<PresenceData> | PresenceData;
+}
+interface VideoContext {
+  video: boolean;
+  duration: number;
+  currentTime: number;
+  paused: boolean;
+}
+function getTimestamps(
+  videoTime: number,
+  videoDuration: number
+): Array<number> {
+  const startTime = Date.now(),
+    endTime = Math.floor(startTime / 1000) - videoTime + videoDuration;
+  return [Math.floor(startTime / 1000), endTime];
+}
+
+const pages: PageContext[] = [
+  {
+    // anime info page
+    middleware: (ref, [video]) =>
+      !!ref.location.pathname.match(/\/anime\/(.*)/gi) && !video,
+    exec: (
+      context,
+      data,
+      { strings }: { strings: { [key: string]: string } }
+    ) => {
+      if (!context) return null;
+      data.state = strings.browsing;
+      data.details = document.title;
+      return data;
+    }
+  },
+  {
+    // watch page
+    middleware: (ref) =>
+      !!ref.location.pathname.match(/^\/(.*)-episode-(\d+)/gi),
+    exec: (
+      context,
+      data,
+      {
+        strings,
+        video
+      }: { strings: { [key: string]: string }; video?: VideoContext }
+    ) => {
+      if (!context) return null;
+      if (video && video.video) {
+        const [start, end] = getTimestamps(
+          Math.floor(video.currentTime),
+          Math.floor(video.duration)
+        );
+        if (!video.paused) {
+          data.state = strings.play;
+          data.startTimestamp = start;
+          data.endTimestamp = end;
+        } else {
+          data.state = strings.pause;
+          delete data.startTimestamp;
+          delete data.endTimestamp;
+        }
+      } else {
+        data.state = strings.browsing;
+        if (data.startTimestamp) delete data.startTimestamp;
+        if (data.endTimestamp) delete data.endTimestamp;
+      }
+      data.details = document.title;
+      return data;
+    }
+  },
+  {
+    middleware: (ref, [video]) => ref && !video,
+    exec: (
+      context,
+      data,
+      { strings }: { strings: { [key: string]: string } }
+    ) => {
+      if (!context) return null;
+      data.details = strings.browsing;
+      return data;
+    }
+  }
+],
+ presence = new Presence({
+  clientId: "778672856347312129"
+});
+
+let currentVideo: VideoContext;
+
+const initialize = async () => {
```

Comment:
for what reason you are initializing because events works every 1 seconds or something :)